File: Level3/5.네트워크/3.cpp
#include <string>
#include <vector>

using namespace std;

int visit[200];

void dfs(vector<vector<int>> &computers, int fr, int to) {
  visit[to] = true;
  for (int i = 0; i < computers[to].size(); i++) {
    if (visit[i] || computers[to][i] == 0) continue;
    dfs(computers, to, i);
  }
}

int solution(int n, vector<vector<int>> computers) {
  int answer = 0;
  for (int i = 0; i < n; i++) {
    if (visit[i]) continue;
    for (int j = 0; j < n; j++) {
      if (computers[i][j] == 0 || visit[j]) continue;
      answer++;
      visit[i] = true;
      dfs(computers, i, j);
    }
  }

  return answer;
}

File: Level3/20.여행경로/3.js
function solution(tickets) {
  const candidates = [];
  const visit = {};

  function dfs(L, path, curr) {
    if (L === tickets.length) {
      candidates.push(path);
      return;
    }

    for (let i = 0; i < tickets.length; i++) {
      const [start, end] = tickets[i];
      if (visit[i] || start !== curr) continue;

      visit[i] = true;
      dfs(L + 1, path + end + ',', end);
      visit[i] = false;
    }
  }

  dfs(0, 'ICN,', 'ICN');

  return candidates.sort()[0].split(',').slice(0, -1);
}
